plugins/devlink/fu-devlink-device.c

```diff
@@ -0,0 +1,580 @@
+		const gchar *name, *value;
+
+		if (mnl_attr_get_type(attr) != DEVLINK_ATTR_INFO_VERSION_FIXED &&
+		    mnl_attr_get_type(attr) != DEVLINK_ATTR_INFO_VERSION_RUNNING &&
+		    mnl_attr_get_type(attr) != DEVLINK_ATTR_INFO_VERSION_STORED)
+			continue;
+
+		if (mnl_attr_parse_nested(attr, fu_devlink_netlink_attr_cb, ver_tb) != MNL_CB_OK)
+			continue;
+
+		if (!ver_tb[DEVLINK_ATTR_INFO_VERSION_NAME] ||
+		    !ver_tb[DEVLINK_ATTR_INFO_VERSION_VALUE])
+			continue;
+
+		name = mnl_attr_get_str(ver_tb[DEVLINK_ATTR_INFO_VERSION_NAME]);
+		value = mnl_attr_get_str(ver_tb[DEVLINK_ATTR_INFO_VERSION_VALUE]);
+
+		version_info = g_hash_table_lookup(version_table, name);
+		if (!version_info) {
+			version_info = g_new0(FuDevlinkVersionInfo, 1);
+			g_hash_table_insert(version_table, (void *)name, version_info);
+		}
+
+		switch (mnl_attr_get_type(attr)) {
+		case DEVLINK_ATTR_INFO_VERSION_FIXED:
+			version_info->fixed = value;
+			break;
+		case DEVLINK_ATTR_INFO_VERSION_RUNNING:
+			version_info->running = value;
+			break;
+		case DEVLINK_ATTR_INFO_VERSION_STORED:
+			version_info->stored = value;
+			break;
+		default:
+			continue;
+		}
+	}
+
+	return version_table;
+}
+
+#define FU_DEVLINK_DEVICE_INSTANCE_ID_PREFIX "DEVLINK\\"
+
+static void
+fu_devlink_device_instance_id_cb(gpointer key, gpointer value, gpointer user_data)
+{
+	FuDevlinkVersionInfo *version_info = value;
+	GString *instance_id = user_data;
+	gchar *name = key;
+
+	if (!version_info->fixed)
+		return;
+	if (g_strcmp0(instance_id->str, FU_DEVLINK_DEVICE_INSTANCE_ID_PREFIX))
+		g_string_append_printf(instance_id, "&");
+	g_string_append_printf(instance_id, "%s_%s", name, version_info->fixed);
+}
+
+static FuDevice *
+fu_devlink_device_get_component(FuDevice *device, const gchar *name)
+{
+	GPtrArray *children = fu_device_get_children(device);
+
+	for (gint i = children->len - 1; i >= 0; i--) {
+		FuDevice *component = g_ptr_array_index(children, i);
+		if (!g_strcmp0(fu_device_get_name(component), name))
+			return g_object_ref(component);
+	}
+	return NULL;
+}
+
+typedef struct {
+	FuDevice *device;
+	GString *instance_id;
+} FuDevlinkDeviceUpdateComponentCtx;
+
+static void
+fu_devlink_device_update_component_cb(gpointer key, gpointer value, gpointer user_data)
+{
+	FuDevlinkDeviceUpdateComponentCtx *ctx = user_data;
+	FuDevlinkVersionInfo *version_info = value;
+	const gchar *name = key;
+	const gchar *version;
+	g_autoptr(FuDevice) component = NULL;
+
+	if (version_info->stored)
+		version = version_info->stored;
+	else if (version_info->running)
+		version = version_info->running;
+	else
+		return;
+
+	component = fu_devlink_device_get_component(ctx->device, name);
+	if (!component) {
+		g_autofree gchar *component_instance_id =
+		    g_strdup_printf("%s&COMPONENT_%s", ctx->instance_id->str, name);
+		component = fu_devlink_component_new(fu_device_get_context(ctx->device),
+						     component_instance_id,
+						     name);
```

Comment:
i think it would be more ideomatic to set the name in the _new() and then add the instance_id from the name in FuDevlinkComponent->probe()

Reply:
Perhaps I misunderstood your comment. The component_ instance_id is something we on-fly generate here and pass to component_new. If it is not stored there in instance_id directly, it would have to be stored elsewhere so the probe could read it. What's the point of that? Note that COMPONENT_%s is just a suffix. On real system, looks something like this: "DEVLINK\DRIVER_mlx5_core&fw.psid_MT_0000000528&COMPONENT_fw"

Reply:
@hughsie not sure you didn't miss my reply.